### PR TITLE
Masking with the real masks

### LINKED ISSUE #1: Update masking function to consider missing inputs
Currently, (truly) missing values are still passed to the model in `train.py`. If a value is missing it should get masked in the masking function and also in the loss. 

- [ ] Masking function accepts a (true) mask as input, and only masks an additional number of tokens
- [ ] Loss ignores this true mask during training
- [ ] Test to make sure this works as expected

One edge case to deal with is if a higher number of true values are masked than the masking ratio - those points might have to be ignored, so we should see how often this is the case.

### PR DESCRIPTION
**Pull request by gabrieltseng**
- [x] Update masking function to consider a "real" mask as well
- [x] Update train script to mask out "real" mask in loss
- [x] Update eval tasks to use the masking too ~~(will likely require a batch size of 1, which could really slow things down). In the future implement masking within the model's attention mechanism.~~ I do sub-batches per unique number of masked tokens in a batch, which I think is a good trade-off.

Closes #1

### CODE REVIEW

**Comment by rubencart** [comment]
Mmm I might be biased by my NLP background but I don't really like the constraint of having to input exactly equally sized sequences. Using attention masks to deal with padding or missing values is very common in all kinds of transformers, e.g. [timeseries](https://huggingface.co/docs/transformers/v4.35.2/en/model_doc/time_series_transformer#transformers.TimeSeriesTransformerModel.forward.past_observed_mask), [vision (object detection)](https://github.com/facebookresearch/detr/blob/3af9fa878e73b6894ce3596450a8d9b89d918ca9/models/transformer.py#L56-L58), [text](https://huggingface.co/docs/transformers/v4.35.2/en/model_doc/roberta#transformers.RobertaModel.forward.attention_mask). And has the advantage that sequences do not need to be of equal length.

I gave implementing it a go: https://github.com/WorldCereal/presto-worldcereal/pull/12. I have to admit at first I thought it was going to be pretty easy but then, since we change the inputs into a sequence of tokens (per channel group etc), it actually proved to be a little bit trickier :-). The PR implements masking a variable number of tokens using attention_masks, only for the evaluation task. For MAE training the create mask function still outputs an equal number of masked tokens per sequence.

For MAE training, now the real missing values count towards the mask ratio, so the number of masked tokens is equal for all samples in a batch. But with attention masks we could choose: Would it make more sense to have the loss applied to an equal fraction of non-missing outputs per sample (hence ignoring missing values for the mask ratio) or would it make more sense to have an equal amount of missing/masked tokens (of which any given number might have a loss applied to it)? I'm leaning more towards the former, but this probably doesn't matter as much, the main point was to allow missing values during downstream tasks :).

You could still order your dataset based on number of missing values for efficiency, since padding just causes unneeded computation, but it would be nice to not _have_ to do that.

File: src/presto.py [hunk 1]
```diff
@@ -452,27 +450,30 @@ def forward(
...
+        upd_mask = torch.cat((torch.zeros(x.shape[0])[:, None].to(device), upd_mask), dim=1)
+        orig_indices = torch.cat(
+            (torch.zeros(x.shape[0])[:, None].to(device).int(), orig_indices + 1),
+            dim=1,
+        )
 
         # apply Transformer blocks
         for blk in self.blocks:
-            x = blk(x)
+            x = blk(x, attn_mask=~upd_mask.bool())
```
Comment (gabrieltseng) [inline src/presto.py:471]: @rubencart, what happens here if I have some values of `x` which are masked?

i.e. `x` has shape `[b, t, 128]` for `t` (mostly) unmasked tokens. What happens if for `b=2`, `t=1` is masked?

I am specifically worried that when we return `self.norm(x.mean(dim=1))` on line 475 below, we will be including tokens which would have otherwise been removed.

Comment (rubencart) [reply src/presto.py:471]: The attention ignores them. But yes we have to exclude them from the mean, good catch
